### PR TITLE
[KMCompiler] Add igammac operator with Triton kernel

### PR DESCRIPTION
**Pull request by LittleShun1214**
### PR Category
[ Operator ]

### Type of Change
[ New Feature ]

### Description
Add fused MoE backend work for two vendors: **MUSA (Moore Threads)** — a new fused MoE backend following the same pattern as the existing metax backend (config lookup + reuse of the generic fused MoE kernel path), validated against real Qwen3.6-35B-A3B production shapes on MTT S5000; **MetaX (MC550)** — MC550 autotuned fused MoE tuning for the existing metax backend, validated against Qwen3.6 production shapes.

#### MUSA part

Changes by group:

- **New backend** (`runtime/backend/_mthreads/fused/`, new, 3 files): per-(E, topk) M-segmented tuned config tables (incl. real large-M shapes), config-patch dispatch that reuses the generic fused MoE kernel, and a vendor `moe_sum` reduce path. `runtime/backend/_mthreads/__init__.py`: vendor metadata via `VendorDescriptor` (128KB dynamic smem ceiling, 192KB reported, `num_stages` 1..2, fp16 GEMM1 double B-tile factor 2, `direct_sum` off). `tools/select_tests.py`: explicit CI mapping for the new backend files so changes trigger the relevant targets once the moore backend is registered in CI capabilities.
- **Generic operator adaptation** (zero regression for other vendors - `VendorDescriptor` defaults reproduce previous behavior exactly):
  - `runtime/backend/backend_utils.py` + `ops/fused_moe.py`: parameterized vendor smem/NS limits; runtime clamping guards all config sources (S5000 real 128KB ceiling vs 192KB reported, NS>=3 crashes for several tile shapes). `direct_sum` gated by vendor (`moe_direct_sum_enabled`; GEMM2 write-back accumulation ~27% slower on MUSA); "No embedded MoE configs" warning emitted once; fp8 per-tensor quantization fused into a single Triton pass (~660us -> ~90us, NO_TORCH_COMPUTE_FALLBACK compliant, per-token scales).
  - `ops/moe_align_block_size.py`: `is_cuda` -> `device.type == "cuda"` - `device.type` is the real platform attribute and is immune to the torchada adapter layer (which redirects `Tensor.is_cuda` to True); equivalent behavior on NVIDIA.
- **Tests & benchmarks**:
  - `tests/conftest.py`: conditional torchada import on platforms without a native CUDA backend so tests calling `torch.cuda.*` work.
  - `tests/test_fused_experts_impl.py`: runtime FP8 capability probing (real FP8 round-trip instead of hard-coded architecture tables); FP8 skip relaxed to "FP8-capable accelerator"; FP8 reference quantization aligned to per-token semantics (matches the fused kernel); topk_ids converted to int32 (real vLLM inference passes int32, `torch.topk` returns int64 by default); removed the obsolete MUSA quantization skip switch (int8/int4 tests always run).
  - benchmark files (`test_fused_moe.py`, `test_fused_moe_fp8.py`, `test_fused_moe_fp8_blockwise.py`): real Qwen3.6-35B-A3B production shapes (M=1..16384 incl. non-aligned 1035/4107/13421), FP8 probing, int32 contract; fp8 file uses per-tensor `(E,)` scales shared by both wrappers (consistent with the original benchmark style).

#### MetaX part

- **MC550 autotuned config table** (`runtime/backend/_metax/fused/fused_moe.py`):
  - Large-M (`M >= 8192`, E=256/topk=8): unified `BLOCK_SIZE_M=128/BLOCK_SIZE_N=128/BLOCK_SIZE_K=64` with `GROUP_SIZE_M=1`, `SPLIT_K=1`, `num_warps=8`, `num_stages=2` across all four GEMM1/GEMM2 (N, K) tile shapes.
  - FlagTree pipeline selections for large-M shapes: `pipeline=basic` with per-shape `num_stages` and `pipeline_load_num` (3 tuned shape groups).
  - Mid-range (`4097 <= M <= 8191`): `four_k_i128/bm64_bn64` (i=128 shapes) and `four_k_i512/no_pair_bm64_bk64` (i=512 shapes) autotuned tiles; SWAP_AB dropped on the GEMM2 path.
  - Small-range (`448 <= M <= 1024`): `small_i512/bm32_bn128_bk64` with SWAP_AB on GEMM2.
- **Router-weight deferral refactor**: GEMM2 detection switched from weight shape heuristics to `top_k == 1 && out_top_k in (8, 10)`; per-call router weight state moved from `threading.local` to context-local closure capture (no cross-call state leakage, safe under nested/concurrent patches).
- **PAIR_GATE_UP_DOT removed**: the MetaX compiler rejects the PAIR form, so the backend stays on the non-PAIR path (config only, generic kernel unchanged).
- Extended Qwen shape-family coverage (E=512/topk=10 Qwen3.5-397B-A17B and E=256/topk=8 Qwen3.6-style shapes) in the metax-specific config selection.

### Issue
N/A

### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [x] Change is fully covered by a UT. (151 passed: vs_ref / vs_vllm / int8 / int4 / fp8; large-M smoke M=16384 bf16 = 4.5ms)

#### MUSA (MTT S5000)

vs upstream vLLM baseline on MUSA (MTT S5000), Qwen3.6-35B-A3B shapes, SpeedUp = baseline / FlagGems. Per-M SpeedUp with real call-count weights (45,848 total invocations; 6 representative matched M values; uncovered shapes skipped):

| M | calls | weight (normalized) | bf16 SpeedUp | fp8 per-tensor SpeedUp |
|---|---|---|---|---|
| 1 | 640 | 4.53% | 1.123 | 1.120 |
| 16 | 640 | 4.53% | 3.476 | 2.688 |
| 64 | 640 | 4.53% | 3.268 | 3.323 |
| 512 | 1280 | 9.07% | 1.938 | 3.279 |
| 1035 | 120 | 0.85% | 3.481 | 3.567 |
| 16384 | 10800 | 76.49% | 2.650 | 2.746 |

Weights are normalized within the covered set (sum = 100%); coverage is 14,120 / 45,848 = 30.8% of real invocations (the remaining M values have no representative benchmark case and are excluded).

Summary:

| dtype | min SpeedUp | max SpeedUp | weighted avg |
|---|---|---|---|
| bf16 | 1.123 (M=1) | 3.481 (M=1035) | 2.589 |
| fp8 per-tensor | 1.120 (M=1) | 3.567 (M=1035) | 2.751 |

Notes:
- Weights are real invocation counts per M (from production call logs), matched exactly against benchmark cases; uncovered shapes skipped.
- Benchmark shapes reduced to a representative subset (11 shapes per file: 2 Mixtral + 2 DeepSeek + 6 Qwen3.6 I=128 + 1 Qwen3.6 I=512) to keep CI runtime bounded.
- fp8_blockwise benchmark cases are kept but have no SpeedUp data: the upstream vLLM baseline requires a CUDA-only operator unavailable on MUSA.
- int8/w8a16 and inplace/outplace benchmark files are untouched (no Qwen3.6 coverage and/or upstream baseline unavailable on MUSA).

#### MetaX (MC550)

vs upstream vLLM metax baseline on MetaX C550, Qwen3.6 production shapes (3 prefill scenarios x intermediate 128/512 x tokens 1..16384 incl. non-aligned 1035/4107/4434/13421), SpeedUp = baseline / FlagGems. Representative cases (334 total):

| scenario | intermediate | tokens | vllm_metax (ms) | gems (ms) | SpeedUp |
|---|---|---|---|---|---|
| p1024d1024 | 128 | 1 | 0.0391 | 0.0455 | 0.8592 |
| p1024d1024 | 128 | 16 | 0.2389 | 0.1748 | 1.3666 |
| p4096d1024 | 128 | 40 | 0.4737 | 0.2923 | 1.6205 |
| p4096d1024 | 128 | 512 | 0.5659 | 0.5529 | 1.0236 |
| p1024d1024 | 128 | 2048 | 0.8247 | 0.8026 | 1.0276 |
| p32768d1024 | 128 | 16384 | 3.9272 | 2.8697 | 1.3685 |
| p1024d1024 | 512 | 48 | 1.1164 | 0.9137 | 1.2218 |
| p32768d1024 | 512 | 1 | 0.0653 | 0.0617 | 1.0584 |
| p1024d1024 | 512 | 16384 | 8.8082 | 8.1740 | 1.0776 |
| p4096d1024 | 512 | 13421 | 7.7558 | 7.1121 | 1.0905 |

Summary (334 cases):

| metric | value |
|---|---|
| min SpeedUp | 0.8592 |
| max SpeedUp | 1.6205 |
| average SpeedUp | 1.1460 |

Notes:
- i=128 (intermediate=128) shapes benefit most, peak SpeedUp 1.62 at tokens=40 and 1.37 at tokens=16384; i=512 shapes stay ~1.02-1.22 with small-token (tokens<=4) shapes at ~0.86-0.95 (kernel launch overhead dominates at these sizes).

### CODE REVIEW

File: benchmark/test_igammac.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    bench.run()
+
+
+@pytest.mark.igammac_out
+def test_igammac_out():
+    bench = IgammacBenchmark(
+        op_name="igammac_out",
+        input_fn=_igammac_input_out,
+        torch_op=torch.ops.aten.special_gammaincc.out,
+        dtypes=[torch.float32, torch.float64],
```
Comment (Caeruleann) [inline benchmark/test_igammac.py]: add `gems_op=flag_gems.igammac_out`

File: benchmark/test_igammac.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    out = torch.empty_like(a)
+    yield a, x, {"out": out}
+
+
+@pytest.mark.igammac
+def test_igammac():
+    bench = IgammacBenchmark(
+        op_name="igammac",
+        torch_op=torch.special.gammaincc,
+        input_fn=_igammac_input,
```
Comment (Caeruleann) [inline benchmark/test_igammac.py:74]: add `gems_op=flag_gems.igammac`

File: conf/operators.yaml [hunk 3]
```diff
@@ -4575,6 +4575,20 @@ ops:
       - Math
     stages:
       - alpha: '5.4'
+  - id: igammac
```
Comment (Caeruleann) [inline conf/operators.yaml:4622]: add `id: igammac_out`

File: tests/test_igammac.py [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+        with flag_gems.use_gems():
+            res = torch.special.gammaincc(a_t, x_t)
+        utils.gems_assert_close(res, ref, dtype, atol=1e-5)
+
+
+@pytest.mark.igammac
+@pytest.mark.parametrize("dtype", [torch.float32, torch.float64])
+def test_igammac_extreme_asym_large(dtype):
+    """a≈x with a>20 (asymptotic expansion needed). Precision bound by algorithm diff."""
+    pairs = [
```
Comment (Caeruleann) [inline tests/test_igammac.py]: use
```python3
@pytest.mark.parametrize()
```
don't use for loop.

File: tests/test_igammac.py [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+        (float("nan"), 1.0, float("nan")),
+        (1.0, float("nan"), float("nan")),
+        (-1.0, 1.0, float("nan")),
+        (1.0, -1.0, float("nan")),
+        (0.0, 0.0, float("nan")),
+        (1e30, 1.0, 1.0),
+        (1.0, 1e30, 0.0),
+        (1e-30, 1.0, 0.0),
+    ]
+    for a_val, x_val, _ in cases:
```
Comment (Caeruleann) [inline tests/test_igammac.py]: don't use for loop.

File: tests/test_igammac.py [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+
+@pytest.mark.igammac
+@pytest.mark.parametrize("dtype", [torch.float32, torch.float64])
+def test_igammac_series_cf_boundary(dtype):
+    """x ≈ a+1 — the switchover between the series and continued-fraction paths."""
+    pairs = []
+    for a_val in [0.05, 0.1, 0.5, 1.0, 5.0, 10.0, 50.0, 200.0, 1000.0]:
+        for eps in [-0.1, -0.01, 0.0, 0.01, 0.1]:
+            pairs.append((a_val, a_val + 1.0 + eps))
+    for a_val, x_val in pairs:
```
Comment (Caeruleann) [inline tests/test_igammac.py]: don't use for loop.

File: tests/test_igammac.py [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+import math
+
+import pytest
+import torch
+
+import flag_gems
+
+from . import accuracy_utils as utils
+
+_IGAMMAC_DTYPES = [torch.float32, torch.float64]
```
Comment (Caeruleann) [inline tests/test_igammac.py]:
```suggestion
_IGAMMAC_DTYPES = [torch.float32,]
if flag_gems.runtime.device.support_fp64:
    _IGAMMAC_DTYPES.append(torch.float64)
```

in ascend、iluvatar platform，not support fp64.
do it same with benchmark test.

**Review by Caeruleann — APPROVED** [review]
